### DISCUSSION TITLE
TR1Main Installer and Configurator

### DISCUSSION THREAD

**Discussion by Richard-L** [LostArtefacts/TRX#402] [General]
I would like to motivate some thought making on what might be desirable UX for TR1M's feature configuration. The project has gotten really far in quite short time, and the "configurability" we already enjoy through JSON editing at this point is pretty advanced. Many things the fans have desired for long are now possible, and continue to be added. As amazing as TR2M is, its big shortcoming to my eyes is that its configuration and initial setup are rather concealed and lacking user guidance, basically requiring a lot of trial and error with those `.reg` files, not to mention file structure. Standalone OpenLara similarly has no indication as to where asset files need to come from and where to place them. This is quite weak UX, and I hope we might make a difference with TR1M, while of course I'm aware that such task isn't a small undertaking – probably the very reason why no other project has done it yet. We had briefly touched upon such installer before: https://github.com/rr-/Tomb1Main/discussions/247#discussioncomment-1678263

My main concern is to generate bigger following for TR1M, hoping to establish it as the main project to turn to when people want to enjoy classic TR1. We would hope the speedrunning community might also take more interest at some point, but this is a task that requires bigger involvement and "political" efforts to get anywhere, even though the project has already begun walking that road with some of the JSON settings aimed at speedruns.

The process of choosing what TR1M has to offer had ideally been a positive and pleasurable experience, rather than a setup chore. Let me suggest a twofold process:

1. Import TR1 assets
Could have two options:
    1. Scan computer for TR1: A GOG/Steam install, PS1/SAT/PC disc inserted physically or mounted virtually.
    2. Drag'n'Drop files into "TR1M-Installer" window. This could literally be any files. People shouldn't have to think what the right files are. The installer should scan anything people throw at it, and then show a list of recognised assets, where the minimum requirement to try TR1M would be `TITLE.PHD` and `LEVEL1.PHD`, as else "New Game" could not be selected (strictly speaking), unless you're fine with people starting from any available level, such as perhaps UB files. For an extreme case, the gym could be greyed out in the menu or just say "unavailable", as its `GYM.PHD` was never made freeware.

    Cf the asset overviews I tried assembling in #247.

    __NB__: Several assets can have _more than one_ version. Fan awareness for this increased only in more recent years. Compare for example Lara's handstand animation in the gym between PS1/PC and Saturn. XProger has made asset formats from all three platforms compatible with OpenLara, which helps greatly to illustrate. First PS1/PC, then Saturn: http://xproger.info/projects/OpenLara/shots/handstand_PC_SAT.mp4

    These are minute differences, but of course we love them. So for a would-be "fulminant" asset importer, it could go as far as to list the different asset versions known to the game, and allow people to pick and choose. I'd imagine this might be something quite popular with TRLE folk, should TR1M allow them to explore more options they didn't know existed to make their levels "special".

2. Configure the game and TR1M features
First to list all of the settings and their current places:
    1. __in-game__: resolution, input mapping, sound volumes, etc
    2. __Tomb1Main.json5__: TR1M features
    3. __Tomb1Main_gameflow__: more TR1M features (essentially)
    4. __Tomb1Main_gameflow_ub__: again

    UX wise, all of these had ideally been in one and the same place, but since the in-game settings are so innate, I suppose they'll just stay.

    For any TR1M features, there had ideally been immediate feedback indicating to the user what an option does. I would suggest something like an interactive preview, where a dummy window of TR1 previews selected options in real-time, next to a large list of options with radio buttons to them. So for example someone could tap through health bar colours and see them previewed, or hear an example of something like "secret chime fix" at hand of audio samples, etc. We would have to find ways to minimise the amount of content to manage here, and ideally let the game itself do the previewing, perhaps at hand of slipstreamed savestates for illustration.

    The resulting configuration choices would then be saved as presets. These could allow something like an "editor's choice" or "recommended" preset, and sets like "PS1 authentic", "DOS authentic", "Steam authentic", etc. This would also be useful in conversation with the TR speedrunning police, as in an ideal world we might create presets for "Glitchless", "Glitched", "No Damage", etc etc together with them, that they officially "approve" to make TR1M legitimate for runs, and could even coauthor and upload themselves.

All of this just to share some well-meaning thoughts. I realise it's quite over-the-top, but as always I'm happy if you consider some of it useful.

**Comment by rr-** [comment]
Personally I'm not interested in coding any of this, but I'm certainly interested in having this :)  I welcome other programmers to take a shot at this and we can talk on how to consolidate. I imagine this could be a separate project linked on the webpage; we could even transform to an organisation here on GH.

**Reply by Richard-L** [reply]
Perfectly understandable ;)

**Comment by Richard-L** [comment]
#149 actually exists which I had forgotten. Linking it here.

**Comment by rr-** [comment]
Installer is now on a review.

**Reply by walkawayy** [reply]
Released in 2.10 :tada:
